ddtrace/internal/ci_visibility/git_client.py

```diff
@@ -0,0 +1,189 @@
+from typing import Tuple  # noqa
+
+import tenacity
+
+from ddtrace.ext.git import build_git_packfiles
+from ddtrace.ext.git import extract_commit_sha
+from ddtrace.ext.git import extract_latest_commits
+from ddtrace.ext.git import extract_remote_url
+from ddtrace.ext.git import get_rev_list_excluding_commits
+from ddtrace.internal.logger import get_logger
+
+from .. import compat
+from ..utils.http import Response
+from ..utils.http import get_connection
+from .constants import AGENTLESS_DEFAULT_SITE
+
+
+log = get_logger(__name__)
+
+RESPONSE = None
+
+
+class CIVisibilityGitClient(object):
+    RETRY_ATTEMPTS = 5
+
+    def __init__(
+        self,
+        api_key,
+        app_key,
+        base_url="",
+    ):
+        # type: (str, str, str) -> None
+        self._serializer = CIVisibilityGitClientSerializerV1(api_key, app_key)
+        self._worker = None  # type: Optional[Process]
+        self._base_url = "https://api.{}".format(os.getenv("DD_SITE", AGENTLESS_DEFAULT_SITE))
+        self._response = RESPONSE
+
+    def start(self, cwd=None):
+        # type: (Optional[str]) -> None
+        if self._worker is None:
+            self._worker = Process(
+                target=CIVisibilityGitClient._run_protocol,
+                args=(self._serializer, self._base_url, self._response),
+                kwargs={"cwd": cwd},
+            )
+            self._worker.start()
+
+    def shutdown(self, timeout=None):
+        # type: (Optional[float]) -> None
+        if self._worker is not None:
+            self._worker.join(timeout)
+            self._worker = None
+
+    @classmethod
+    def _run_protocol(cls, serializer, base_url, _response, cwd=None):
+        # type: (CIVisibilityGitClientSerializerV1, str, Optional[Response], Optional[str]) -> None
+        repo_url = cls._get_repository_url(cwd=cwd)
+        latest_commits = cls._get_latest_commits(cwd=cwd)
+        backend_commits = cls._search_commits(base_url, repo_url, latest_commits, serializer, _response)
+        rev_list = cls._get_filtered_revisions(backend_commits, cwd=cwd)
+        if rev_list:
+            with cls._build_packfiles(rev_list, cwd=cwd) as packfiles_prefix:
+                cls._upload_packfiles(base_url, repo_url, packfiles_prefix, serializer, _response, cwd=cwd)
+
+    @classmethod
+    def _get_repository_url(cls, cwd=None):
+        # type: (Optional[str]) -> str
+        return extract_remote_url(cwd=cwd)
+
+    @classmethod
+    def _get_latest_commits(cls, cwd=None):
+        # type: (Optional[str]) -> list[str]
+        return extract_latest_commits(cwd=cwd)
+
+    @classmethod
+    def _search_commits(cls, base_url, repo_url, latest_commits, serializer, _response):
+        # type: (str, str, list[str], CIVisibilityGitClientSerializerV1, Optional[Response]) -> list[str]
+        payload = serializer.search_commits_encode(repo_url, latest_commits)
+        response = _response or cls.retry_request(base_url, "/search_commits", payload, serializer)
```

Comment:
It's used by tests to generate a dummy response. It has to be done this way as opposed to mocking because some of the tests run this function in a subprocess, and Mock objects are not picklable.